chore(internal-release): 0.12.0

sfoster1 opened the pull request:
This is the tracking branch for internal release 0.12.0.

sfoster1 committed:
chore: release notes for hotfix 6.3.1 (#12788)

Release notes for the hotfix detailing bugfixes and improvements.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

sanni-t committed:
fix(api, app): thermocycler hold time on API v2.14 (#12814)

* added holdTimeSeconds param to TC setTargetBlockTemperature, wired up to TC module context in API
* updated setTargetBlockTemperature in v7 schema
* updated app run log command text for setTargetBlockTemperature & waitForBlockTemperature
---------

Co-authored-by: Brian Cooper <brian@opentrons.com>

TamarZanzouri committed:
feature(api, hardware-controller): allow hardware to move both mounts at the same time (#12844)

jbleon95 committed:
feat(shared-data): adapter definitions and stacking overlaps (#12811)

Updated shared-data labware schema, labware definitions that can be stacked on adapters, and added adapter definitions

koji committed:
fix(app): fix robot system version screen and modal (#12849)

update robot system version screen and robot system version modal to align with hi-fi design

shlokamin committed:
Merge branch 'release' into edge

SyntaxColoring committed:
feat(api): Enforce apiLevel 2.15 for Flex-style deck slot names (#12846)

jgbowser committed:
fix(app): Protocol Run DQA updates (#12845)

*fix(app): Protocol Run DQA updates
Fixes some issues noted during design qa on the protocol run screens of the ODD

vegano1 committed:
fix(api): set the firmware binary dir for the Flex to /usr/lib/firmware (#12866)

ahiuchingau committed:
feat(hardware): add limit switch backoff to home sequence (#12824)

ryanthecoder committed:
fix(hardware): don't try to send set_current to brushed motor (#12869)

* don't set current for gripper_g

* == not =

koji committed:
fix(app): fix white screen when tapping view error details button (#12875)

When errors are undefined or no error data, the app makes the button disabled

andySigler committed:
feat(hardware-testing): Dispense liquid-class settings are configurable per test volume (#12867)

* dispense liquid-class settings configurable per test volume

* fix incorrect volume for 8ch P50

* updates leading-air-gaps to be latest-greatest

sfoster1 committed:
chore: remove codecov failure noise (#12880)


overly sensitive and desensitize us to red X's causing us to miss actual
failures.

Make the codecov checker much much more lenient.

vegano1 committed:
fix(api): revert firmware binaries dir change and use /usr/bin/firmware for Flex (#12881)

sfoster1 committed:
chore(robot-server): restore deleted tests (#12883)

These were removed accidentally by #12880

sfoster1 committed:
chore(robot-server): fix timing-out test (#12879)

The ot2-only tests were running an ot3 test and that simply won't do.
The right fix for this is more in depth but this will fix it quickly.

sfoster1 committed:
feat: add HTTP firmware update (#12731)

This PR is an extremely large refactor of all three of the primary python packages that run on a robot to enable firmware update of connected microcontrollers via HTTP, and to reduce the amount of canbus traffic that only exists to poll what instruments are attached.

It does all of this by adding a new first-class concept to the hardware management of the robot: the subsystem. The subsystem is an enum like the axes, and it represents a self-contained program running on a processor logically or physically distinct from the robot server (so you can't talk to it via normal linux commands, basically). In practice, that just means the microcontrollers on the flex; there are pretty easy extensions to have this also handle the motor controller board on the OT-2, and in fact I added an entry to the enum for just that purpose.

The subsystems have a firmware version, and an ok status, and can be updated. Internally, a subsystem being "not ok" really means "stuck in the bootloader"; this is a reasonable thing to say of a subsystem, because it's a program. It wouldn't be ok to say of an instrument, because in that case you wouldn't be able to know any of the information we need about a pipette: what its serial is, what its model is, things like that.

Subsystems are handled in hardware by a new object that handles their lifecycles and update cycles and manages discovering them, including asynchronously, so that interfaces above this don't need to actively poll.

We need a way for clients to figure out that they need to interact with the subsystems API, and so now when instruments that need a firmware update - whether that's because they're out of date, or because they're not booting properly - we return a BadPipette or BadGripper from GET /instruments, distinguishable by the presence of ok: false in the response (ok: true means the response is as before). This bad response also gives a route to the appropriate subsystem handler (see below).

Finally, this PR makes some new routes for clients to update subsystems - mostly instruments. These routes combine a way to view ongoing updates and a way to view historical updates. Ongoing updates are findable by subsystem, but when they complete they disappear; historical updates are viewable by key, and stick around while the robot is booted.

The client workflow for a client starting an update here would be:

-    See that an update is needed by checking GET /subsystems/status
-    Start an update by POST /subsystems/updates/:subsystem, and get an ID
-    Check on the update by GET /subsystems/updates/all/:id

The client workflow for a client seeing that somebody else started an update would be:

-    See that an update is going by checking GET /subsystems/updates/current
-    Get its status by checking GET /subsystems/updates/all/:id

If two clients race to start an update, that's fine because the loser gets redirected to the ongoing update.

## Major changelog:

* refactor(api): ThreadManager handles async gens

Async generator functions are different from normal async functions, in the same
way that generator functions are different from normal functions. When called,
they return a special async generator object. This is a problem in
ThreadManager land, because those async generator objects need to be
interacted with via `await anext(gen)` in the same loop that created
them.

So, this commit adds a new case in thread manager for "what if you call
an async generator function in the hardware controller". In that case,
what happens is
- We return a proxy async generator function, which is defined in (and
therefore runs in) the calling loop
- We generate a future that is a coroutine running in the hardware
controller loop that captures a queue, and pulls items out of the async
generator in that loop and puts them in a queue
- Our proxy pulls stuff out of the queue and yields this

This is of course extra frustrating because all queue interaction must
take place in the same loop, which must be the loop the queue was
created in. That means that for the queue that moves generator output
from the hardware thread/loop to the calling thread/loop, it has to be
- Created in the hardware thread/loop (via run_coroutine_threadsafe and
a wrapper)
- Inserted into in the hardware thread/loop (fine, this happens in the
coroutine that drains the generator)
- Gotten from in the hardware thread/loop, so every await queue.get()
needs to happen in a run_coroutine_threadsafe()

It all works! But it's pretty annoying.

* fix(opentrons_hardware): errors during fw update

We weren't ever checking any exceptions that got raised during the
run-update task. We can do that by getting the result of the future
afterward (and throwing it away - the only thing we care about is that
result() will raise any exception that happened in the task).

* refactor(robot-server): global IDNotFound error

We have these scattered hither and yon, let's get it in one place and
start reusing it.

* refactor(api,robot-server): make SubSystem exposed

We had this type called OT3SubSystem, which was a representation of the
subsystems (programmable devices, really) that could be bolted on to the
machine. It was reasonably useful but never used to its full extent.

Now, let's use it. First, we're going to want it to work on the OT-2 as
well (just for completeness) so let's remove the OT-3 part of the name
and add a way to get to it from a mount; we'll leave a temporary alias
for other code.

As an example of how we'll use it, we can make a new set of models in
the server under subsystems/.

* feat(robot-server): expose instrument subsystems

Add a new element "subsystem" to the instrument responses that lets
clients link an instrument to a subsystem. The subsystem is the
representation of the programmable microcontroller on the machine, and
is the part you can update. This is an important distinction to make
because if a subsystem isn't working properly, you generally can't
identify it as an instrument; but you can always identify it as a
subsystem. The subsystem retrieved here will be useful later in some new
routes.

Note: this commit definitely won't pass tests and mypy or like work on
its own, because this is a big interconnected PR.

* refactor(api): remove fw update info from instrs

These were going to be used to trigger instrument firmware updates, but
we're going to base all that on subsystems now so they can be removed.

* refactor(hardware): remove pipette lookup fallback

It's now been long enough since we got updated bootloaders that we don't
need the fallback for pipettes with bootloaders from EVT; let's remove
the fallback for looking up the pipette type in the bootloader.

Note: This commit does not stand alone because the callsites for it
aren't modified in the commit (because they're very shortly to be gone).
It's set apart so it stands out as a thing that changed.

* refactor(hardware): put fw update in an object

This looks (and is) extremely awkward, but it lets us decoy this object
instead of having to hang mocks off weird strings, and in that context
it's smooth as silk.

* refactor(hardware): network exposes ok status

We're going to do all this subsystem management, but a subsystem can
appear whether the device is in its application or its bootloader. We
need to expose that information somewhere, so let's start at the bottom
and use the network submodule from opentrons hardware to detect whether
the device is OK or not, and put that information in the device
information it exposes.

* refactor(hardware): split tool detect and resolve

We have this ToolDetector thing. It listens for messages from the head
and returns the kinds of pipettes and gripper that are attached, polling
those devices for their information.

Let's refactor that to split those two things into a detect (which is
fast, and involves no further communication) and a resolve (which
involves lots more further communication). This lets us have a nice
async iterator for the detection and a lovely resolution mechanism that
can be called for each detect event, and then use all that stuff
precisely once for the oneshot.

* refactor(hardware-control): add subsystem manager

If we want to hook a bunch of stuff off subsystems, we need a place to
handle them. We want to control the lifecycle of subsystems and their
updates, and that means we should get a lovely object to do this in.

This object has three jobs:
1. Manage the discovery of the "normal" subsystems on the network (i.e.
not stuff that's normally appearing and disappearing like tools), which
is what the ot3controller.probe_network() method used to do; this will
now use the opentrons_hardware.hardware_control.network
2. Manage those subsystems that do appear and disappear, and get the
additional information that they have. This uses the tools detector
infrastructure that we recently changed, triggered off the outcome of
network probes, and exposes and fuses all that information and turns it
into subsystem data.
3. Update those subsystems, including managing multiple ongoing updates
from multiple callers and handling the detection lifecycles of
subsystems as they get updated.

The primary interface here are the properties, because a key feature is
that this manages lifecycles _on its own_. You don't have to ask to
detect tools, because it's triggered off the head microcontroller
notifications; you don't have to re-refresh after an update, it will do
that.

There are some fun implementation details in here; for instance we use
the ever-exotic condition variable to allow for a bit of controlled
state that represents both whether we've completed a tool detection
cycle and whether the worker task got an error.

This commit doesn't stand alone because of the changes in the types
file; a followup will change how the ot3controller and hardware api work.

* refactor(api): use subsystems as primary objects

Now that we have a subsystem manager, we can use it by turning the
hardware api into considering _subsystems_ as a new first-class object
alongside instruments.

subsystems are what get updated, not instruments or mounts; all those
subsystems being ready to run are what defines the api being ready to
operate, not several separate checks. There's not a lot of new
functionality here, mostly things get removed because they have
been added in new submodules.

Because we now rely on submodules as a first class object and subsystems
can be instruments, we can turn the instrument detection mechanism into
something that queries what subsystems are currently present - a flow
which does not require any canbus traffic and will always be correct
because tools are updated asynchronously from the head.

* refactor(robot-server): add firmware updater

The firmware update manager will be the primary functionality driver of
the new subsystems endpoints. It handles the following needs:

- Clients should be able to see what needs updating
- Clients should be able to update things that need updating
- Clients should be able to make decisions based on updates that were
  started by other clients

The way to do this is by having a two-tier system of process management.
Processes can be either ongoing or completed. While processes are
ongoing, you can access them by the subsystem that they're updating.
Because hardware will raise exceptions if you try and start an update
process for a subsystem that's already being updated, there can only
ever be one ongoing update per subsystem.

When an update completes, whether it succeeds or fails, it is no longer
ongoing and goes over to be stored for the long term (until reboot) by
the id it was created with, and can later be retrieved by that ID.

The process flow for a client would be something like:
1. See that a subsystem needs updating
2. Begin an update
   2a. Either the update is in process, in which case we can get its
   status by subsystem, including its ID, or
   2b. The update begins and we can get its status and ID by subsystem
3. Watch the update by its ID, which will always be valid

Because 3 is by-id, you can't miss an event like the update ending while
you had your laptop closed or something. Because you can access an
update process initially by subsystem, you can display the status of an
update process that was started by someone else.

There's some fun implementation details here, like the use of handles to
make sure that callers can only interact with the ongoing process by one
end of a queue. The idea is anything outside this module either
interacts with a manager singleton or with process handles it gets from
the monitor singleton.

* feat(robot-server): add subsystem management

The new subsystem management endpoints allow clients to see the status,
start the firmware updates, and monitor past and present firmware
updates for the microcontrollers attached to the robot system.

The status routes are at GET /subsystems/status and GET /subsystem/status/:subsystem.
These can be used to see what subsystems are present; and whether each
subsystem is OK. While the acceptable values for GET
/subsystem/:subsystems and the subsystem names returned from GET
/subsystem are technically controlled, it's probably a better idea for
clients to get the names from GET /subsystems/status for later use.

The update monitoring routes are at GET /subsystems/updates/current, GET
/subsystems/updates/current/:subsystem, GET /subsystems/updates/all and
GET /subsystems/updates/all/:id. The difference between current and all
is that curreent is keyed by subsystem and only contains resources for
updates that are running at this precise instant; all contains all
updates since the robot booted, ongoing or done.

Updates can be started at POST /subsystems/updates/:subsystem. Starting
an update here has one of three major outcomes:
1. Everything is fine and the update is started; you get the ID for an
update process, and you get a 201 that points you to the ongoing update
2. There's already an update in progress for this subsystem; you get a
303 that points you to the update for that subsystem
3. The subsystem doesn't need an update, or the subsystem isn't present,
or something like this and you get an error and no subsystem.

If you get an update, you can now monitor its status. Clients should
use either "trying to start an update" or getting an ongoing update for
discovery; for monitoring the update, they should use the update ID.
This is because when an update completes it is no longer ongoing and
therefore no longer accessible via the current updates, so if you miss
the completion it will just disappear. Via ID, it will stick around and
you get a complete status.

* refactor(api,robot-server): Update fw in server

When we start the robot, we need to run any required firmware updates -
maybe we just got a system update, maybe somebody pushed something,
maybe the system is booting with a new instrument - and we want those
updates to be visible via HTTP. That means they need to be update
processes, and that means that they need to happen via the firmware
update manager in the server.

So let's hoist the initial update out of the hardware controller build
function and up into the robot server system startup tasks, which
requires refactoring them a bit to handle the new ot3-only behavior.

As a side-effect of this refactor, we should no longer be trying to
blink a nonexistent button on the OT-3, and we should have the server
become ready for interaction _much_ faster since it no longer waits for
firmware updates to be complete before sending its systemd ready notice.
While this might cause a weird time in the ODD interaction while
firmware updates are ongoing but the app seems fine, the solution to
that is the frontend changes to view and handle updates.

* refactor(robot-server): get_ot3_hardware fixups

FastAPI actually evaluates the type hints on dependency functions at
runtime. This is a Really Big Problem for us because we can't rely on
OT3API being importable. It means that get_ot3_hardware can't be a
dependency; that also means it can't use dependencies and callers have
to pass in actual objects.

Also, you can't really do isinstance(some_thread_manager, OT3API)
because it's a thread manager not an OT3API; add a passthrough wrapper
for the thread manager that lets you ask what it's wrapping.


* feat(robot-server): subsystems for bad instrs

The whole point of the subsystems interface is to have something to
present when an instrument is physically present but not actually able
to operate as an instrument for whatever reason. Since the need for the
subsystems is based on the need to handle instruments, and clients want
to be polling instruments, have /instruments return information about
the subsystem if an instrument needs an update or isn't ok so that
clients need to go look at the subsystems.

* refactor(robot-server): clean up instr responses

The instrument responses in the bad vs good case were wildly different
which makes for a bad experience with clients. They can now be
distinguished via the value of the `ok` response element - bad is bad,
good is good. "OK" here means "fully ready to 

* refactor(api): update at most passed arguments

We would unconditionally update anything that needed updating, which is
not what we want; instead, we want to update the intersection of what
needs updating and what was requested to be updated.

* refactor(robot-server): split hw postinit

By putting postinit in a separate task, we can let the server come up
ready earlier. This lets us avoid issues with timeouts and allow clients
to see any details of the postinit tasks that we might want to expose -
for instance, the status of ongoing first-boot firmware updates for
connected subsystems.

* hardware: handle USB comm errors

If we try and probe while the CANbus has an update going on, that's
fine, those devices just don't respond. But if we try and probe the USB
while the rear panel is updating, we can't find the usb device and will
get an exception. Since updates do a network probe when they complete,
if we have multiple concurrent updates then if the USB update doesn't
complete first the others will try and fail to probe it.

But really, all a communication error via USB means is that the probe
failed - that device is not present. So let's treat it like that.

* refactor(hardware,api): probe_specific

When we probe the network, we send a broadcast ping and wait for
anything. We're askign the network in general what's present.

But with the subsystem manager, in a lot of situations what we really
want is to ask whether _specific devices_ are present, since we're
maintaining an estimate of what's present before then. We only want to
do the full broadcast when we start.

So add a new method called probe_specific, which instead of broadcasting
and returning _anything_ (and yelling at you if it misses what you
expect) will broadcast and look _only for responses from what you tell
it_, and not touch the rest of cached state.

This can then be used when a firmware update completes to prevent the
pings from looking for other devices in the middle of their own updates.


* refactor: check devices more leniently after update

When we update, we check that all the devices that should have just
completed an update have come back up and are present on the network.
This lets us verify that the update succeeded. But they're rebooting
after an update, so we need to give them a little longer to start with a
nice retry system.

Also, add a mechanism to indicate that specific tools are no longer present.

* refactor: exclusive can access for restarts

There's a period when our microcontroller applications and bootloaders
start where their can peripherals are running and accepting messages,
but no filters have been loaded. This usually isn't a problem, but it's
a really big problem when it's the bootloader, and at the same time
there's an update running and sending data for a different node - in
that case, the starting device gets a message for the other one and
reads it in. This doesn't cause corruption because we're checking
checksums and total packet counts, but it does make the update fail.

To fix that, we can introduce the concept of exclusive access to the can
messenger, since can messengers are shared. This pretty much boils down
to a lock. send() and ensure_send() take the lock all the time; but also
you can take the lock explicitly and prevent use of the canbus for some
longer time, and then use exclusive_send and exclusive_ensure_send to
avoid deadlocks.

By holding exclusive access in the updater while we're telling devices
to enter the bootloader, we should prevent any udpate data packets from
hitting the bus in the vulnerable period while a device is restarting.

There's some improvements that can be made here - there's no reason to
prevent concurrent access to the canbus when nothing's holding the
exclusive lock - but this works for now.

* refactor: update progress fix

The progress needs to be intified and multiplied in hardware control,
since that's where we turn things into ints.

* refactor: timeout tweak

It's now pretty important to be on the nose about noticing something has
rebooted since we have exclusive access so make shorter timeouts and
more retries


This reverts commit 53b8a9d3448aafd00f556f0c8868a4b6b23e15b6.

koji committed:
fix(app): software keyboard dqa (#12876)

Update software keyboard layout, height and colors. Also export customDisplay variable as a constant
since CustomKeyboard and NormalKeyboard use the same display method.

fsinapi committed:
fix(api): use consistent CriticalPoints when moving to maintenance position (#12878)

Use the MOUNT critical point for moving to X and Y position, but still calculate final Z position based on the tip of whatever's attached to the mount

vegano1 committed:
feat(api,robot-server): expose module calibration offsets via /modules endpoint. (#12860)

fsinapi committed:
feat(api, robot-server): add setting to disable the status bar (#12887)

* Added feature flag to enable/disable status bar

* Added function to enable/disable the status bar, and call it whenever the feature flags are edited

* Added tests for enable/disable status bar

* Added tests for status bar settings

smb2268 committed:
feat(app, shared-data): use new lower position for mounting plate attachment (#12787)

fix RLIQ-399

andySigler committed:
feat(hardware-testing): Add ot2 photometric script to hardware-testing (#12896)

koji committed:
fix(shared-data, app): update the slot name for Flex (#12892)

Update the slot name from 7+10 to A1+B1 for thermocycler

koji committed:
optimize images (#12868)

optimize images that are added for Flex

jerader committed:
refactor(app): 96 channel assets and instruments dashboard updates (#12807)

closes RLIQ-405
Co-authored-by: smb2268 <sarah.breen@opentrons.com>

sfoster1 committed:
fix(robot-server): Better logging for postinit (#12895)

We have this awkward situation around logging, particularly for
exceptions, in hardware postinit. We really want to be able to print
stuff like where an exception occurred, but this is tragically really
annoying to do without being in the correct exception stack. There's a
way around this but it's weird: in the original exception stack, create
a special TracebackException instance, which will capture the right
things theoretically. However, that class _is not an exception_, and so
either everything has to return an exception or we have to wrap it in an
exception so we can raise it and use the old methods... so that's what
we do.

koji committed:
feat(app): add LocationIcon (#12888)

Add LocationIcon component

ryanthecoder committed:
feat(api, shared-data): add pipette plunger backlash compensation (#12858)

vegano1 committed:
refactor(api,robot-server): change module calibration slot usage from Int to coordinate string. (#12889)

koji committed:
feat(app): update device reset slideout to align with the design (#12874)

Update device reset slideout to align with the design

koji committed:
fix(app): fix instrument details dqa and protocol setup button dqa (#12872)

Fix instrument details dqa and protocol setup button dqa

koji committed:
fix(app): fix modal dqa round3 (#12884)

Fix modal dqa round3

sfoster1 committed:
chore: internal-release -> edge for 0.11.0

shlokamin committed:
feat(app): wire up led lights in ODD robot settings (#12904)

jerader committed:
feat(app): flex pipette static assets (#12906)

closes RLIQ-411

sfoster1 committed:
chore: release notes for 0.12.0

TamarZanzouri reviewed (approved):
my changes look good! thank you!

sfoster1 committed:
fix(api): increase pipette detection timeout (#12916)

We have a timeout for getting the pipette info response during tool
attach, and it was too short, breaking the pipette attach process. By
increasing the timeout we fix the issue.

Closes RQA-961

Laura-Danielle committed:
refactor(robot-server): use the server.log file to store logs coming in from the robot server (#12894)

ahiuchingau committed:
fix(api): home plunger & mount axis properly after drop tip (#12857)

sfoster1 committed:
fix(hardware): allow longer for pipettes to boot (#12941)

Pipettes sometimes don't boot fast enough on attach and miss the
InstrumentInfoRequest that happens when the head tells us a new
instrument is present. This didn't used to be an issue because we
wouldn't send the request until the user was finished screwing things
in, so we just didn't notice.

In addition to our previous timeout - or, at a certain level, instead of
it - where we'd wait a long time for the pipette to respond, we'll also
now retry sending the request a couple times too.

Fixes RQA-998

ahiuchingau committed:
refactor(api): Update flex acc speeds based on EE recs (#12946)

fsinapi committed:
fix(hardware): don't stop reading from CAN after most exceptions (#12963)
